File: src/main/kotlin/auth/Token.kt
package devctl.auth


import kotlinx.serialization.*
import java.time.Instant

@Serializable
data class Token(

    @SerialName("access_token")
    val accessToken: String,

    @SerialName("refresh_token")
    val refreshToken: String? = null,

    @SerialName("id_token")
    val idToken: String,

    @SerialName("token_type")
    val tokenType: String,

    @SerialName("expires_in")
    val expiresIn: Long,

    @SerialName("created_at")
    val createdAt: Long = Instant.now().epochSecond,

    val scope: String
)

fun Token.IsExpired(): Boolean {
    return Instant.now().epochSecond > expiresIn + createdAt
}

File: auth/token.go
package auth

import (
	"encoding/json"
	"time"
)

type Token struct {
	AccessToken  string `json:"access_token"`
	RefreshToken string `json:"refresh_token"`
	IDToken      string `json:"id_token"`
	TokenType    string `json:"token_type"`
	ExpiresIn    int64  `json:"expires_in"`
	CreatedAt    int64  `json:"created_at"`
}

func UnmarshalToken(data []byte) (Token, error) {
	var r Token
	err := json.Unmarshal(data, &r)
	return r, err
}

func (r *Token) Marshal() ([]byte, error) {
	return json.Marshal(r)
}

func (t *Token) IsExpired() bool {
	return t.CreatedAt + t.ExpiresIn < time.Now().Unix() - 60
}
